ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Iterator.hs

```diff
@@ -257,24 +258,49 @@ newIterator itEnv@IteratorEnv{..} getItEnv registry from to = do
           => ExceptT (UnknownRange blk) m (DeserialisableIterator m blk)
     start = lift (ImmDB.getTipInfo itImmDB) >>= \case
       Origin                          -> findPathInVolDB
-      At (tipSlot, tipHash, _tipIsEBB) ->
+      At (tipSlot, tipHash, tipIsEBB) ->
         case atSlot endPoint `compare` tipSlot of
           -- The end point is < the tip of the ImmutableDB
           LT -> streamFromImmDB
           EQ | withHash endPoint == tipHash
                 -- The end point == the tip of the ImmutableDB
              -> streamFromImmDB
-             | otherwise
-                -- The end point /= the tip of the ImmutableDB. Either the
-                -- fork is too old, or the end point (or the tip of the
-                -- ImmutableDB) is an EBB. For example, the tip of the
-                -- ImmutableDB is an EBB and the end point is the regular
-                -- block after it.
+
+             -- The end point /= the tip of the ImmutableDB.
+             --
+             -- There are three possibilities:
+             --
+             -- 1. The end point is the regular block (or a different EBB?)
+             --    with the same slot number as the EBB which is currently at
+             --    the tip of the ImmutableDB. In this case, we must look in
+             --    the VolatileDB for the block, as we know it is not in the
+             --    ImmutableDB.
+             --
+             -- 2. The end point is the EBB with the same slot number as the
+             --    regular block which is currently at the tip of the
+             --    ImmutableDB. In this case, we can stream from just the
+             --    ImmutableDB, as the end point is in the ImmutableDB.
+             --
+             -- 3. The fork is too old, i.e., the end point refers to a block
+             --    (or EBB) not on the current chain, but with the same slot
+             --    number as our tip.
```

Comment:
If I understand correctly, I think it would be clearer to explicitly state here that Case 3 should always raise the `ForkTooOld` exception.

(I'm not certain I understand correctly, because I'm not certain what exactly you mean by "current chain" here: if the ImmDB tip is an EBB U, it's possible that multiple proper blocks V, V', V'', ... with the same slot as U are in the VolDB and we should be able to open an iterator for any of them, regardless of which one is on the "current(ly selected) chain", right?)

Reply:
New comment is great.